tfhe/src/core_crypto/experimental/algorithms/lwe_shrinking_keyswitch_key_generation.rs

```diff
@@ -0,0 +1,106 @@
+//! Module containing primitives pertaining to [`LWE shrinking keyswitch keys
+//! generation`](`crate::core_crypto::entities::LweKeyswitchKey#key-switching-key`).
+
+use crate::core_crypto::algorithms::generate_lwe_keyswitch_key;
+use crate::core_crypto::commons::generators::EncryptionRandomGenerator;
+use crate::core_crypto::commons::math::random::{Distribution, Uniform};
+use crate::core_crypto::commons::parameters::*;
+use crate::core_crypto::commons::traits::*;
+use crate::core_crypto::entities::LweSecretKey;
+use crate::core_crypto::experimental::commons::parameters::LweSecretKeySharedCoefCount;
+use crate::core_crypto::experimental::entities::{
+    LweShrinkingKeyswitchKey, LweShrinkingKeyswitchKeyOwned,
+};
+
+pub fn generate_lwe_shrinking_keyswitch_key<
+    Scalar,
+    NoiseDistribution,
+    InputKeyCont,
+    OutputKeyCont,
+    KSKeyCont,
+    Gen,
+>(
+    input_lwe_sk: &LweSecretKey<InputKeyCont>,
+    output_lwe_sk: &LweSecretKey<OutputKeyCont>,
+    lwe_keyswitch_key: &mut LweShrinkingKeyswitchKey<KSKeyCont>,
+    noise_distribution: NoiseDistribution,
+    generator: &mut EncryptionRandomGenerator<Gen>,
+) where
+    Scalar: Encryptable<Uniform, NoiseDistribution>,
+    NoiseDistribution: Distribution,
+    InputKeyCont: Container<Element = Scalar>,
+    OutputKeyCont: Container<Element = Scalar>,
+    KSKeyCont: ContainerMut<Element = Scalar>,
+    Gen: ByteRandomGenerator,
+{
+    assert!(
+        lwe_keyswitch_key.output_key_lwe_dimension() == output_lwe_sk.lwe_dimension(),
+        "The destination LweKeyswitchKey output LweDimension is not equal \
+        to the output LweSecretKey LweDimension. Destination: {:?}, output: {:?}",
+        lwe_keyswitch_key.output_key_lwe_dimension(),
+        output_lwe_sk.lwe_dimension()
+    );
+
```

Comment:
Assert 
- `shared_randomness < *_lwe_sk.len()`
- `shrunk_input_lwe_secret_key.len() == lwe_keyswitch_key.input_key_lwe_dimension()`
- `input_lwe_sk[0..shared_randomness] == output_lwe_sk[0..shared_randomness]`

Reply:
- likely useful indeed
- 2 actually checked by the generate lwe ksk so will remove the assert at that level
- we don't check key content generally speaking so not sure for the last one

Reply:
1 is checked by the shrinking keyswitch constructor